docker/pypi/wmagent/bin/manage-common.sh

```diff
@@ -346,150 +357,94 @@ _parse_wmasecrets(){
-    local MATCH_ACDC_URL=`cat $WMA_SECRETS_FILE | grep ACDC_URL | sed s/ACDC_URL=//`
-    local MATCH_DBS3_URL=`cat $WMA_SECRETS_FILE | grep DBS3_URL | sed s/DBS3_URL=//`
-    local MATCH_DQM_URL=`cat $WMA_SECRETS_FILE | grep DQM_URL | sed s/DQM_URL=//`
-    local MATCH_REQUESTCOUCH_URL=`cat $WMA_SECRETS_FILE | grep REQUESTCOUCH_URL | sed s/REQUESTCOUCH_URL=//`
-    local MATCH_CENTRAL_LOGDB_URL=`cat $WMA_SECRETS_FILE | grep CENTRAL_LOGDB_URL | sed s/CENTRAL_LOGDB_URL=//`
-    local MATCH_WMARCHIVE_URL=`cat $WMA_SECRETS_FILE | grep WMARCHIVE_URL | sed s/WMARCHIVE_URL=//`
-    local MATCH_AMQ_CREDENTIALS=`cat $WMA_SECRETS_FILE | grep AMQ_CREDENTIALS | sed s/AMQ_CREDENTIALS=//`
-    local MATCH_RUCIO_HOST=`cat $WMA_SECRETS_FILE | grep RUCIO_HOST | sed s/RUCIO_HOST=//`
-    local MATCH_RUCIO_AUTH=`cat $WMA_SECRETS_FILE | grep RUCIO_AUTH | sed s/RUCIO_AUTH=//`
-    local MATCH_RUCIO_ACCOUNT=`cat $WMA_SECRETS_FILE | grep RUCIO_ACCOUNT | sed s/RUCIO_ACCOUNT=//`
-    local MATCH_TEAMNAME=`cat $WMA_SECRETS_FILE | grep TEAMNAME | sed s/TEAMNAME=//`
-    local MATCH_AGENT_NUMBER=`cat $WMA_SECRETS_FILE | grep AGENT_NUMBER | sed s/AGENT_NUMBER=//`
-
-
-    # database settings (mysql or oracle)
-    if [ "x$MATCH_ORACLE_USER" == "x" ]; then
+    # Auxiliary function to parse WMAgent.secrets or MariaDB.secrets files
+    # and load a set of variables from them
+    # :param $1: Path to WMAgent.secrets or file (Default: $WMA_SECRETS_FILE)
+    # :param $2: String with variable names to be checked (Default: *)
+    # :retrun:   Error value if one or more values have been left unset in the secrets file
+    local errVal=0
+    local value=""
+    local secretsFile=${1:-$WMA_SECRETS_FILE}
+    local varsToLoad=${2:-""}
+
+    [[ -f $secretsFile ]] || {
+        echo "$FUNCNAME: ERROR: Password file: $secretsFile does not exist"
+        echo "$FUNCNAME: ERROR: Either set WMA_SECRETS_FILE environment variable to a valid file or check that $HOME/WMAgent.secrets exists"
+        return $(false)
+    }
+
+    # If no list of variables to be loaded was given assume all of them.
+    # Building the list by parsing the secrets file itself.
+    [[ -n $varsToLoad ]] || {
+       varsToLoad=`grep -v "^[[:blank:]]#" $secretsFile  |grep \= | awk -F\= '{print $1}'`
+    }
+
+    # Here we validate every variable for itself before loading it
+    for varName in $varsToLoad
+    do
+        _parse_wmasecrets $secretsFile $varName || {
+            let errVal+=1
+            echo "$FUNCNAME: ERROR: Bad value found for $varName"
+            return $errVal
+        }
+    done
+
+    # Now load them all
+    for varName in $varsToLoad
+    do
+        value=`grep -E "^[[:blank:]]*$varName=" $secretsFile | sed "s/ *$varName=//"`
+        [[ $varName =~ ^RESOURCE_ ]] && declare -g -A $varName
+        eval $varName=$value
+        [[ -n $varName ]] || { echo "$FUNCNAME: ERROR: Empty value for: $varName=$value"; let errVal+=1 ;}
```

Comment:
Given that we are already validating the secrets parameter a few lines above:
https://github.com/dmwm/CMSKubernetes/pull/1462/files#diff-01c5d9be01b6e6ea0ee4b12d5262b4a641bc3bdc55a79853ca90abb819996f96R386

isn't this L399 redundant?

Reply:
no, it is not
because 
* between the moment when you set the value and the moment when you "initially" parse and validate the the whole secrets file  there are  few steps including an extra declaration which may cause an error in the `eval` step due to an obscured value put for those `RESOURCE_*`  (which we  cannot predict what it could be) 
* This is also a precaution, because at some point, if in the future we decide to add  yet another functionality to be parsed, and we work only in the `_parse_wmasecrets` file (s it should be), but we somehow make a mistake and create a bug, we will be extra protected from this line here to avoid empty values loaded.

Reply:
I would leave this as it is

Reply:
Yes, I don't think it's worth it to change this PR only because of that line.

I still fail to understand the logic for checking these variables. Isn't this line:
```
[[ ${value,,} =~ $badValuesReg ]] && { echo "$FUNCNAME: ERROR: Bad value for: $var=$value"; let errVal+=1 ;}
```
from `_parse_wmasecrets` already catching the case where a variable is left empty?

I am not sure I understand your first bullet. Are you saying that someone might mistype one of the RESOURCE_ variables, which would then fail the `        eval $varName=$value` command?

Reply:
hi @amaltaro 
> I am not sure I understand your first bullet. Are you saying that someone might mistype one of the RESOURCE_ variables, which would then fail the         eval $varName=$value command?

You got it correctly. With one minor detail. The value may not mandatory be mistyped, but it could be someone actually tried to configure those resources in the form of a regular array, rather than an associative array. e.g. `RESOURCE_OPP3=("T3_US_UCSD" 3000  5000 "normal")` instead of `RESOURCE_OPP3=([name]="T3_US_UCSD" [run]=3000  [pend]=5000 [state]="normal")` This will result in the following error in the evaluation command:

```
:~$ declare -A RESOURCE_OPP3
:~$ RESOURCE_OPP3=("T3_US_UCSD" 3000  5000 "normal")
-bash: RESOURCE_OPP3: "T3_US_UCSD": must use subscript when assigning associative array
-bash: RESOURCE_OPP3: 3000: must use subscript when assigning associative array
-bash: RESOURCE_OPP3: 5000: must use subscript when assigning associative array
-bash: RESOURCE_OPP3: "normal": must use subscript when assigning associative array
:~$ echo ${RESOURCE_OPP3[*]}

```   

Which results in an empty value. While the expected behavior is:

```
:~$ declare -A RESOURCE_OPP3
:~$ RESOURCE_OPP3=([name]="T3_US_UCSD" [run]=3000  [pend]=5000 [state]="normal")
:~$ echo ${RESOURCE_OPP3[*]}
normal 3000 T3_US_UCSD 5000
```
Since the `_parse_wmasecrets` function does not have a section for detailed validation of value  types other than the generic value presence and bad regular expression... this cannot be caught by this function. Hence the presence of the extra check.